File: Week-Wise-Assignments/Assignment 7/Systematic Solutions/C Programs/4_merge_sort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Function to display array
void displayArray(int arr[], int n, const char* message) {
    printf("%s: ", message);
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

// Global variables for tracking operations
int mergeOperations = 0;
int recursiveDepth = 0;
int maxDepth = 0;

// Algorithm for Merge operation:
// 1. Create temporary arrays for left and right subarrays
// 2. Copy data to temporary arrays
// 3. Merge temporary arrays back into original array in sorted order
// 4. Compare elements from both arrays and pick smaller one
void merge(int arr[], int left, int mid, int right, int showSteps) {
    int n1 = mid - left + 1;
    int n2 = right - mid;
    
    // Create temporary arrays
    int* leftArr = (int*)malloc(n1 * sizeof(int));
    int* rightArr = (int*)malloc(n2 * sizeof(int));
    
    // Copy data to temporary arrays
    for (int i = 0; i < n1; i++) {
        leftArr[i] = arr[left + i];
    }
    for (int j = 0; j < n2; j++) {
        rightArr[j] = arr[mid + 1 + j];
    }
    
    if (showSteps) {
        printf("  Merging subarrays [%d, %d] and [%d, %d]\n", left, mid, mid + 1, right);
        printf("    Left subarray:  ");
        for (int i = 0; i < n1; i++) {
            printf("%d ", leftArr[i]);
        }
        printf("\n");
        printf("    Right subarray: ");
        for (int j = 0; j < n2; j++) {
            printf("%d ", rightArr[j]);
        }
        printf("\n");
    }
    
    // Merge the temporary arrays back into arr[left..right]
    int i = 0, j = 0, k = left;
    
    while (i < n1 && j < n2) {
        if (leftArr[i] <= rightArr[j]) {
            arr[k] = leftArr[i];
            if (showSteps) {
                printf("    Comparing %d <= %d: Choose %d\n", leftArr[i], rightArr[j], leftArr[i]);
            }
            i++;
        } else {
            arr[k] = rightArr[j];
            if (showSteps) {
                printf("    Comparing %d > %d: Choose %d\n", leftArr[i], rightArr[j], rightArr[j]);
            }
            j++;
        }
        k++;
    }
    
    // Copy remaining elements of leftArr[], if any
    while (i < n1) {
        arr[k] = leftArr[i];
        if (showSteps) {
            printf("    Copying remaining from left: %d\n", leftArr[i]);
        }
        i++;
        k++;
    }
    
    // Copy remaining elements of rightArr[], if any
    while (j < n2) {
        arr[k] = rightArr[j];
        if (showSteps) {
            printf("    Copying remaining from right: %d\n", rightArr[j]);
        }
        j++;
        k++;
    }
    
    if (showSteps) {
        printf("    Merged result: ");
        for (int idx = left; idx <= right; idx++) {
            printf("%d ", arr[idx]);
        }
        printf("\n\n");
    }
    
    mergeOperations++;
    free(leftArr);
    free(rightArr);
}

// Algorithm for Merge Sort:
// 1. Divide the array into two halves
// 2. Recursively sort both halves
// 3. Merge the sorted halves
void mergeSort(int arr[], int left, int right, int showSteps) {
    if (left < right) {
        recursiveDepth++;
        if (recursiveDepth > maxDepth) {
            maxDepth = recursiveDepth;
        }
        
        int mid = left + (right - left) / 2;
        
        if (showSteps) {
            printf("Dividing array [%d, %d] at mid = %d\n", left, right, mid);
            printf("  Left half: [%d, %d], Right half: [%d, %d]\n", 
                   left, mid, mid + 1, right);
        }
        
        // Sort first and second halves
        mergeSort(arr, left, mid, showSteps);
        mergeSort(arr, mid + 1, right, showSteps);
        
        // Merge the sorted halves
        merge(arr, left, mid, right, showSteps);
        
        recursiveDepth--;
    }
}

// Iterative merge sort implementation
void mergeSortIterative(int arr[], int n) {
    printf("\n=== Iterative Merge Sort ===\n");
    displayArray(arr, n, "Initial array");
    
    int mergeCount = 0;
    
    // Current size of subarrays to be merged
    for (int currSize = 1; currSize < n; currSize = 2 * currSize) {
        printf("\nMerging subarrays of size %d:\n", currSize);
        
        // Pick starting point of left sub array
        for (int leftStart = 0; leftStart < n - 1; leftStart += 2 * currSize) {
            // Calculate mid and right points
            int mid = leftStart + currSize - 1;
            int rightEnd = (leftStart + 2 * currSize - 1 < n - 1) ? 
                          leftStart + 2 * currSize - 1 : n - 1;
            
            if (mid < rightEnd) {
                printf("  Merging [%d, %d] with [%d, %d]\n", 
                       leftStart, mid, mid + 1, rightEnd);
                merge(arr, leftStart, mid, rightEnd, 0);
                mergeCount++;
            }
        }
        
        printf("  After merging size %d: ", currSize);
        displayArray(arr, n, "");
    }
    
    printf("\nIterative Merge Sort completed!\n");
    printf("Total merge operations: %d\n", mergeCount);
}

// Performance analysis function
void performanceAnalysis() {
    printf("\n=== Performance Analysis ===\n");
    
    int sizes[] = {8, 16, 32, 64, 128};
    int numSizes = sizeof(sizes) / sizeof(sizes[0]);
    
    printf("Array Size | Merge Ops | Max Depth | Time Complexity\n");
    printf("-----------|-----------|-----------|----------------\n");
    
    for (int i = 0; i < numSizes; i++) {
        int n = sizes[i];
        int* arr = (int*)malloc(n * sizeof(int));
        
        // Generate random array
        srand(time(NULL) + i);
        for (int j = 0; j < n; j++) {
            arr[j] = rand() % 1000;
        }
        
        // Reset counters
        mergeOperations = 0;
        recursiveDepth = 0;
        maxDepth = 0;
        
        // Perform merge sort
        mergeSort(arr, 0, n - 1, 0);
        
        printf("%10d | %9d | %9d | O(n log n)\n", 
               n, mergeOperations, maxDepth);
        
        free(arr);
    }
}

// Function to demonstrate merge operation
void demonstrateMerge() {
    printf("\n=== Merge Operation Demonstration ===\n");
    
    int arr[] = {2, 5, 8, 1, 3, 9};
    int n = sizeof(arr) / sizeof(arr[0]);
    
    printf("Demonstrating merge of two sorted subarrays:\n");
    displayArray(arr, n, "Array with two sorted parts");
    printf("Left sorted part: [0, 2] = {2, 5, 8}\n");
    printf("Right sorted part: [3, 5] = {1, 3, 9}\n\n");
    
    merge(arr, 0, 2, 5, 1);
    
    displayArray(arr, n, "After merging");
}

// Function to generate different types of arrays
void generateTestArray(int arr[], int n, int type) {
    switch (type) {
        case 1: // Random array
            srand(time(NULL));
            for (int i = 0; i < n; i++) {
                arr[i] = rand() % 100 + 1;
            }
            break;
            
        case 2: // Already sorted array
            for (int i = 0; i < n; i++) {
                arr[i] = i + 1;
            }
            break;
            
        case 3: // Reverse sorted array
            for (int i = 0; i < n; i++) {
                arr[i] = n - i;
            }
            break;
            
        case 4: // Array with duplicates
            srand(time(NULL));
            for (int i = 0; i < n; i++) {
                arr[i] = rand() % 10 + 1;
            }
            break;
    }
}

// Function to check if array is sorted
int isSorted(int arr[], int n) {
    for (int i = 0; i < n - 1; i++) {
        if (arr[i] > arr[i + 1]) {
            return 0;
        }
    }
    return 1;
}

int main() {
    int choice, n;
    
    printf("=== Merge Sort Implementation ===\n");
    printf("Merge Sort: Divide and conquer sorting algorithm\n");
    printf("Time Complexity: O(n log n) for all cases\n");
    printf("Space Complexity: O(n) - Requires additional space for merging\n");
    printf("Stability: Stable (maintains relative order of equal elements)\n\n");
    
    while (1) {
        printf("\n--- Menu ---\n");
        printf("1. Sort custom array (recursive, step-by-step)\n");
        printf("2. Sort custom array (recursive, fast)\n");
        printf("3. Sort using iterative merge sort\n");
        printf("4. Sort test arrays (different types)\n");
        printf("5. Demonstrate merge operation\n");
        printf("6. Performance analysis\n");
        printf("7. Algorithm explanation\n");
        printf("8. Exit\n");
        printf("Enter your choice: ");
        scanf("%d", &choice);
        
        switch (choice) {
            case 1: {
                printf("Enter array size: ");
                scanf("%d", &n);
                
                if (n <= 0 || n > 50) {
                    printf("Invalid size! Please enter size between 1 and 50.\n");
                    break;
                }
                
                int* arr = (int*)malloc(n * sizeof(int));
                
                printf("Enter %d elements: ", n);
                for (int i = 0; i < n; i++) {
                    scanf("%d", &arr[i]);
                }
                
                printf("\n=== Recursive Merge Sort (Step-by-step) ===\n");
                displayArray(arr, n, "Initial array");
                
                // Reset counters
                mergeOperations = 0;
                recursiveDepth = 0;
                maxDepth = 0;
                
                mergeSort(arr, 0, n - 1, 1);
                
                printf("Merge Sort completed!\n");
                printf("Statistics: %d merge operations, max recursion depth: %d\n", 
                       mergeOperations, maxDepth);
                displayArray(arr, n, "Final sorted array");
                
                printf("\nVerification: Array is %s\n", 
                       isSorted(arr, n) ? "correctly sorted" : "NOT sorted");
                
                free(arr);
                break;
            }
            
            case 2: {
                printf("Enter array size: ");
                scanf("%d", &n);
                
                if (n <= 0 || n > 1000) {
                    printf("Invalid size! Please enter size between 1 and 1000.\n");
                    break;
                }
                
                int* arr = (int*)malloc(n * sizeof(int));
                
                printf("Enter %d elements: ", n);
                for (int i = 0; i < n; i++) {
                    scanf("%d", &arr[i]);
                }
                
                printf("\n=== Recursive Merge Sort (Fast) ===\n");
                displayArray(arr, n, "Initial array");
                
                // Reset counters
                mergeOperations = 0;
                recursiveDepth = 0;
                maxDepth = 0;
                
                mergeSort(arr, 0, n - 1, 0);
                
                printf("Merge Sort completed!\n");
                printf("Statistics: %d merge operations, max recursion depth: %d\n", 
                       mergeOperations, maxDepth);
                displayArray(arr, n, "Final sorted array");
                
                free(arr);
                break;
            }
            
            case 3: {
                printf("Enter array size: ");
                scanf("%d", &n);
                
                if (n <= 0 || n > 100) {
                    printf("Invalid size! Please enter size between 1 and 100.\n");
                    break;
                }
                
                int* arr = (int*)malloc(n * sizeof(int));
                
                printf("Enter %d elements: ", n);
                for (int i = 0; i < n; i++) {
                    scanf("%d", &arr[i]);
                }
                
                mergeSortIterative(arr, n);
                free(arr);
                break;
            }
            
            case 4: {
                printf("Enter array size: ");
                scanf("%d", &n);
                
                if (n <= 0 || n > 100) {
                    printf("Invalid size! Please enter size between 1 and 100.\n");
                    break;
                }
                
                printf("\nSelect array type:\n");
                printf("1. Random array\n");
                printf("2. Already sorted array\n");
                printf("3. Reverse sorted array\n");
                printf("4. Array with duplicates\n");
                printf("Enter choice: ");
                
                int type;
                scanf("%d", &type);
                
                if (type < 1 || type > 4) {
                    printf("Invalid choice!\n");
                    break;
                }
                
                int* arr = (int*)malloc(n * sizeof(int));
                generateTestArray(arr, n, type);
                
                const char* typeNames[] = {"", "Random", "Sorted", "Reverse Sorted", "With Duplicates"};
                printf("\nGenerated %s Array:\n", typeNames[type]);
                displayArray(arr, n, "Before sorting");
                
                // Reset counters
                mergeOperations = 0;
                recursiveDepth = 0;
                maxDepth = 0;
                
                mergeSort(arr, 0, n - 1, 0);
                
                printf("Statistics: %d merge operations, max depth: %d\n", 
                       mergeOperations, maxDepth);
                displayArray(arr, n, "After sorting");
                
                free(arr);
                break;
            }
            
            case 5:
                demonstrateMerge();
                break;
                
            case 6:
                performanceAnalysis();
                break;
                
            case 7:
                printf("\n=== Merge Sort Algorithm Explanation ===\n");
                printf("How it works:\n");
                printf("1. Divide: Split the array into two halves\n");
                printf("2. Conquer: Recursively sort both halves\n");
                printf("3. Combine: Merge the sorted halves into a single sorted array\n\n");
                
                printf("Characteristics:\n");
                printf("• Divide and conquer algorithm\n");
                printf("• Stable sorting algorithm\n");
                printf("• Not in-place (requires O(n) extra space)\n");
                printf("• Consistent performance regardless of input\n");
                printf("• Parallelizable (can be implemented in parallel)\n\n");
                
                printf("Time Complexity:\n");
                printf("• Best Case: O(n log n)\n");
                printf("• Average Case: O(n log n)\n");
                printf("• Worst Case: O(n log n)\n");
                printf("• Always performs the same number of operations\n\n");
                
                printf("Space Complexity:\n");
                printf("• O(n) for the temporary arrays used in merging\n");
                printf("• O(log n) for the recursion stack\n");
                printf("• Total: O(n)\n\n");
                
                printf("Variants:\n");
                printf("• Iterative Merge Sort: Avoids recursion overhead\n");
                printf("• In-place Merge Sort: Reduces space complexity\n");
                printf("• Natural Merge Sort: Takes advantage of existing runs\n\n");
                
                printf("When to use:\n");
                printf("• Large datasets where O(n log n) is required\n");
                printf("• When stable sorting is needed\n");
                printf("• External sorting (sorting data that doesn't fit in memory)\n");
                printf("• When consistent performance is important\n");
                break;
                
            case 8:
                printf("Exiting program...\n");
                exit(0);
                
            default:
                printf("Invalid choice! Please try again.\n");
        }
    }
    
    return 0;
}

File: Week-Wise-Assignments/Assignment 7/Systematic Solutions/Python Programs/4_merge_sort.py
def merge(arr, left, mid, right, show_steps=False):
    """
    Algorithm for Merge operation:
    1. Create temporary arrays for left and right subarrays
    2. Copy data to temporary arrays
    3. Merge temporary arrays back into original array in sorted order
    4. Compare elements from both arrays and pick smaller one
    """
    global merge_operations
    
    # Create temporary arrays
    left_arr = arr[left:mid + 1]
    right_arr = arr[mid + 1:right + 1]
    
    if show_steps:
        print(f"  Merging subarrays [{left}, {mid}] and [{mid + 1}, {right}]")
        print(f"    Left subarray:  {left_arr}")
        print(f"    Right subarray: {right_arr}")
    
    # Merge the temporary arrays back into arr[left..right]
    i = j = 0
    k = left
    
    while i < len(left_arr) and j < len(right_arr):
        if left_arr[i] <= right_arr[j]:
            arr[k] = left_arr[i]
            if show_steps:
                print(f"    Comparing {left_arr[i]} <= {right_arr[j]}: Choose {left_arr[i]}")
            i += 1
        else:
            arr[k] = right_arr[j]
            if show_steps:
                print(f"    Comparing {left_arr[i]} > {right_arr[j]}: Choose {right_arr[j]}")
            j += 1
        k += 1
    
    # Copy remaining elements of left_arr, if any
    while i < len(left_arr):
        arr[k] = left_arr[i]
        if show_steps:
            print(f"    Copying remaining from left: {left_arr[i]}")
        i += 1
        k += 1
    
    # Copy remaining elements of right_arr, if any
    while j < len(right_arr):
        arr[k] = right_arr[j]
        if show_steps:
            print(f"    Copying remaining from right: {right_arr[j]}")
        j += 1
        k += 1
    
    if show_steps:
        print(f"    Merged result: {arr[left:right + 1]}")
        print()
    
    merge_operations += 1

def merge_sort(arr, left, right, show_steps=False):
    """
    Algorithm for Merge Sort:
    1. Divide the array into two halves
    2. Recursively sort both halves
    3. Merge the sorted halves
    """
    global recursive_depth, max_depth
    
    if left < right:
        recursive_depth += 1
        if recursive_depth > max_depth:
            max_depth = recursive_depth
        
        mid = left + (right - left) // 2
        
        if show_steps:
            print(f"Dividing array [{left}, {right}] at mid = {mid}")
            print(f"  Left half: [{left}, {mid}], Right half: [{mid + 1}, {right}]")
        
        # Sort first and second halves
        merge_sort(arr, left, mid, show_steps)
        merge_sort(arr, mid + 1, right, show_steps)
        
        # Merge the sorted halves
        merge(arr, left, mid, right, show_steps)
        
        recursive_depth -= 1

def merge_sort_iterative(arr):
    """Iterative merge sort implementation"""
    n = len(arr)
    print(f"\n=== Iterative Merge Sort ===")
    print(f"Initial array: {arr}")
    
    merge_count = 0
    
    # Current size of subarrays to be merged
    curr_size = 1
    while curr_size < n:
        print(f"\nMerging subarrays of size {curr_size}:")
        
        # Pick starting point of left sub array
        left_start = 0
        while left_start < n - 1:
            # Calculate mid and right points
            mid = min(left_start + curr_size - 1, n - 1)
            right_end = min(left_start + 2 * curr_size - 1, n - 1)
            
            if mid < right_end:
                print(f"  Merging [{left_start}, {mid}] with [{mid + 1}, {right_end}]")
                merge(arr, left_start, mid, right_end, False)
                merge_count += 1
            
            left_start += 2 * curr_size
        
        print(f"  After merging size {curr_size}: {arr}")
        curr_size *= 2
    
    print(f"\nIterative Merge Sort completed!")
    print(f"Total merge operations: {merge_count}")
    
    return arr

def performance_analysis():
    """Analyze performance for different array sizes"""
    print(f"\n=== Performance Analysis ===")
    
    sizes = [8, 16, 32, 64, 128]
    
    print("Array Size | Merge Ops | Max Depth | Time Complexity")
    print("-----------|-----------|-----------|----------------")
    
    for size in sizes:
        # Generate random array
        import random
        arr = [random.randint(1, 1000) for _ in range(size)]
        
        # Reset counters
        global merge_operations, recursive_depth, max_depth
        merge_operations = 0
        recursive_depth = 0
        max_depth = 0
        
        # Perform merge sort
        merge_sort(arr, 0, size - 1, False)
        
        print(f"{size:10d} | {merge_operations:9d} | {max_depth:9d} | O(n log n)")

def demonstrate_merge():
    """Demonstrate merge operation"""
    print(f"\n=== Merge Operation Demonstration ===")
    
    arr = [2, 5, 8, 1, 3, 9]
    
    print("Demonstrating merge of two sorted subarrays:")
    print(f"Array with two sorted parts: {arr}")
    print("Left sorted part: [0, 2] = {2, 5, 8}")
    print("Right sorted part: [3, 5] = {1, 3, 9}")
    print()
    
    merge(arr, 0, 2, 5, True)
    
    print(f"After merging: {arr}")

def generate_test_array(size, array_type):
    """Generate different types of test arrays"""
    import random
    
    if array_type == 1:  # Random array
        return [random.randint(1, 100) for _ in range(size)]
    elif array_type == 2:  # Already sorted
        return list(range(1, size + 1))
    elif array_type == 3:  # Reverse sorted
        return list(range(size, 0, -1))
    elif array_type == 4:  # Array with duplicates
        return [random.randint(1, 10) for _ in range(size)]
    else:
        return [random.randint(1, 100) for _ in range(size)]

def is_sorted(arr):
    """Check if array is sorted"""
    return all(arr[i] <= arr[i + 1] for i in range(len(arr) - 1))

# Global variables for tracking operations
merge_operations = 0
recursive_depth = 0
max_depth = 0

def main():
    """Main function with interactive menu"""
    print("=== Merge Sort Implementation ===")
    print("Merge Sort: Divide and conquer sorting algorithm")
    print("Time Complexity: O(n log n) for all cases")
    print("Space Complexity: O(n) - Requires additional space for merging")
    print("Stability: Stable (maintains relative order of equal elements)\n")
    
    while True:
        print("\n--- Menu ---")
        print("1. Sort custom array (recursive, step-by-step)")
        print("2. Sort custom array (recursive, fast)")
        print("3. Sort using iterative merge sort")
        print("4. Sort test arrays (different types)")
        print("5. Demonstrate merge operation")
        print("6. Performance analysis")
        print("7. Algorithm explanation")
        print("8. Exit")
        
        try:
            choice = int(input("Enter your choice: "))
            
            if choice == 1:
                size = int(input("Enter array size: "))
                
                if size <= 0 or size > 50:
                    print("Invalid size! Please enter size between 1 and 50.")
                    continue
                
                print(f"Enter {size} elements separated by spaces: ")
                arr = list(map(int, input().split()))
                
                if len(arr) != size:
                    print("Number of elements doesn't match the specified size!")
                    continue
                
                print(f"\n=== Recursive Merge Sort (Step-by-step) ===")
                print(f"Initial array: {arr}")
                
                # Reset counters
                global merge_operations, recursive_depth, max_depth
                merge_operations = 0
                recursive_depth = 0
                max_depth = 0
                
                merge_sort(arr, 0, len(arr) - 1, True)
                
                print("Merge Sort completed!")
                print(f"Statistics: {merge_operations} merge operations, max recursion depth: {max_depth}")
                print(f"Final sorted array: {arr}")
                
                print(f"\nVerification: Array is {'correctly sorted' if is_sorted(arr) else 'NOT sorted'}")
                
            elif choice == 2:
                size = int(input("Enter array size: "))
                
                if size <= 0 or size > 1000:
                    print("Invalid size! Please enter size between 1 and 1000.")
                    continue
                
                print(f"Enter {size} elements separated by spaces: ")
                arr = list(map(int, input().split()))
                
                if len(arr) != size:
                    print("Number of elements doesn't match the specified size!")
                    continue
                
                print(f"\n=== Recursive Merge Sort (Fast) ===")
                print(f"Initial array: {arr}")
                
                # Reset counters
                merge_operations = 0
                recursive_depth = 0
                max_depth = 0
                
                merge_sort(arr, 0, len(arr) - 1, False)
                
                print("Merge Sort completed!")
                print(f"Statistics: {merge_operations} merge operations, max recursion depth: {max_depth}")
                print(f"Final sorted array: {arr}")
                
            elif choice == 3:
                size = int(input("Enter array size: "))
                
                if size <= 0 or size > 100:
                    print("Invalid size! Please enter size between 1 and 100.")
                    continue
                
                print(f"Enter {size} elements separated by spaces: ")
                arr = list(map(int, input().split()))
                
                if len(arr) != size:
                    print("Number of elements doesn't match the specified size!")
                    continue
                
                merge_sort_iterative(arr)
                
            elif choice == 4:
                size = int(input("Enter array size: "))
                
                if size <= 0 or size > 100:
                    print("Invalid size! Please enter size between 1 and 100.")
                    continue
                
                print("\nSelect array type:")
                print("1. Random array")
                print("2. Already sorted array")
                print("3. Reverse sorted array")
                print("4. Array with duplicates")
                
                array_type = int(input("Enter choice: "))
                
                if array_type < 1 or array_type > 4:
                    print("Invalid choice!")
                    continue
                
                arr = generate_test_array(size, array_type)
                
                type_names = ["", "Random", "Sorted", "Reverse Sorted", "With Duplicates"]
                print(f"\nGenerated {type_names[array_type]} Array:")
                print(f"Before sorting: {arr}")
                
                # Reset counters
                merge_operations = 0
                recursive_depth = 0
                max_depth = 0
                
                merge_sort(arr, 0, len(arr) - 1, False)
                
                print(f"Statistics: {merge_operations} merge operations, max depth: {max_depth}")
                print(f"After sorting: {arr}")
                
            elif choice == 5:
                demonstrate_merge()
                
            elif choice == 6:
                performance_analysis()
                
            elif choice == 7:
                print(f"\n=== Merge Sort Algorithm Explanation ===")
                print("How it works:")
                print("1. Divide: Split the array into two halves")
                print("2. Conquer: Recursively sort both halves")
                print("3. Combine: Merge the sorted halves into a single sorted array\n")
                
                print("Characteristics:")
                print("• Divide and conquer algorithm")
                print("• Stable sorting algorithm")
                print("• Not in-place (requires O(n) extra space)")
                print("• Consistent performance regardless of input")
                print("• Parallelizable (can be implemented in parallel)\n")
                
                print("Time Complexity:")
                print("• Best Case: O(n log n)")
                print("• Average Case: O(n log n)")
                print("• Worst Case: O(n log n)")
                print("• Always performs the same number of operations\n")
                
                print("Space Complexity:")
                print("• O(n) for the temporary arrays used in merging")
                print("• O(log n) for the recursion stack")
                print("• Total: O(n)\n")
                
                print("Variants:")
                print("• Iterative Merge Sort: Avoids recursion overhead")
                print("• In-place Merge Sort: Reduces space complexity")
                print("• Natural Merge Sort: Takes advantage of existing runs\n")
                
                print("When to use:")
                print("• Large datasets where O(n log n) is required")
                print("• When stable sorting is needed")
                print("• External sorting (sorting data that doesn't fit in memory)")
                print("• When consistent performance is important")
                
            elif choice == 8:
                print("Exiting program...")
                break
                
            else:
                print("Invalid choice! Please try again.")
                
        except ValueError:
            print("Invalid input! Please enter a valid number.")
        except Exception as e:
            print(f"An error occurred: {e}")

if __name__ == "__main__":
    main()